docs/internal/abi.md

```diff
@@ -0,0 +1,409 @@
+Objects can also be passed to functions or inserted into arrays, and this ABI defines how that works so that readers and writers, callers and callees all agree on the same representation.
+
+The following ABI proposal is under construction and will change over time.
+The public ABI (when interfacing with foreign languages or exported symbols) will be stable at some point and guaranteed by the compiler.
+But the compiler reserves the right to change the internal ABI at any time, where it can guarantee the changes are not visible from user code.
+
+## Concrete Types
+
+### Machine-word types
+
+Machine-word types, (ex, numeric) are singleton types (no fields) and do not have a _standard_ object representation (see below).
+They are treated special by the compiler and are represented as their machine equivalent bit-widths.
+All numeric types are aligned naturally, except booleans, which align to its storage type.
+
+* `Bool`: bit pattern 0 or 1. The storage is usually 1 byte but can be different.
+* `U8`, `U16`, ... `U128`: power-of-two-bits, unsigned, on the platform's native endian.
+* `I8`, `I16`, ... `I128`: power-of-two-bits, 2's-complement, on the platform's native endian.
+* `F32`, `F64`: IEEE-754 (binary) floating point numbers.
+
+Booleans can be packed (ex. 1 byte = 8 boolean values) for arrays on specific optimisations.
+
+There are discussions on introducing non-power-of-two integers and other floating point sizes, but they won't be present for the first iteration of the compiler.
+
+### Pointers
+
+Pointers have different sizes depending on the target architecture:
+* 32-bit integers representing an address.
+* 64-bit integers representing an address.
+* 128-bit CHERI capabilities granting access to an address range.
+
+All Verona objects are at least 8-byte aligned and so the low three bits of a pointer will always be zero.
+When targeting a system that uses integer addresses, all pointers have the same representation.
+When targeting a CHERI system, the following additional rules apply:
+* Pointers to `cown`s are sealed with a type reserved for the runtime.
+  This ensures that `cown`s in immutable objects cannot be dereferenced by unsafe (sandboxed) code and provides some defense in depth against compiler bugs that would accidentally dereference a `cown` pointer.
+* Pointers to immutable objects are not represented by read-only capabilities.
+  Doing so would require expensive operations for reference-count manipulation.
+  When a pointer to an immutable object is passed via the foreign-code layer, it must have the load-storable-capabilities permission removed, such that no reachable capability will ever provide write permission.
+
+### Classes and interfaces
+
+Concrete classes and interfaces each have their own unique header, with a unique descriptor.
+Following a pointer and decoding the header gives you the type of the object and therefore there is no need to treat interfaces and classes differently.
```

Comment:
Are pointer to object going to be at the beginning of the object's data (so we read the header backwards) or to the header directly (so all element pointers are +2)?